File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.py
import sys
sys.setrecursionlimit(99999999)

N = int(input())

forest = [list(map(int, input().split())) for _ in range(N)]

dx = [0,0,-1,1]
dy = [1,-1,0,0]

dp = [[0 for _ in range(N)] for _ in range(N)]


def dfs(x , y):
    
    if dp[x][y] != 0:
        return dp[x][y]
    
    dp[x][y] = 1
    
    for i in range(4):
        nx , ny = x + dx[i] , y + dy[i]
        if 0 <= nx < N and 0 <= ny < N:
            if forest[x][y] < forest[nx][ny]:
                dp[x][y] = max(dfs(nx, ny) + 1 , dp[x][y])
    return dp[x][y]

answer = 0
for i in range(N):
    for j in range(N):
        answer = max(dfs(i, j) ,answer)
        
print(answer)

File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.js
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n");

const n = Number(input[0]);
const forest = [];

for (let i = 1; i <= n; i++) {
  forest.push(input[i].split(" ").map(Number));
}

const dp = Array.from({ length: n }, () => Array(n).fill(0));

const dx = [1, -1, 0, 0];
const dy = [0, 0, 1, -1];

function dfs(x, y) {
  if (dp[x][y] != 0) {
    return dp[x][y];
  }

  dp[x][y] = 1;

  for (let i = 0; i < 4; i++) {
    const nx = x + dx[i];
    const ny = y + dy[i];

    if (nx >= 0 && nx < n && ny >= 0 && ny < n) {
      if (forest[x][y] < forest[nx][ny]) {
        dp[x][y] = Math.max(dp[x][y], dfs(nx, ny) + 1);
      }
    }
  }

  return dp[x][y];
}

let answer = 0;
for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    answer = Math.max(answer, dfs(i, j));
  }
}

console.log(answer);